minetest/minetest
# Light space perspective shadow mapping

This PR adds Light Space Perspective Shadow Mapping based on [work of Michael Wimmer et al.](https://www.cg.tuwien.ac.at/research/vr/lispsm/shadows_egsr2004_revised.pdf).

The goal is to improve the quality of shadow maps by using more of the texture surface while keeping single-pass shadow generation process. The approach also removes the distortion causing 'curved' SM texels.

The intent of the PR is to improve the quality of shadows specifically at Very Low and Low presets that use low-resolution SM texture and, combined with #11747, resolve or improve #11762. It also resolves #11358 at Very Low preset and improves  #11743 for more quality presets.

## LiSPSM

LiSPSM combines Uniform and Perspective Shadow Maps by manipulating the light frustum to match the perspective of player's camera as close as possible in order to increase the density of SM texels close to the player.

Strength of perspective transformation is controlled by parameter n, which is calculated from the player's camera parameters and then blended with 'infinity' (1/1E5) based on angle between player's view and light direction.

## To do

This PR is Ready for Review.

items in the PR:
* [x] Implement light frustum calculation
* [x] Remove perspective distortion from SM shaders and shadow sampling
* [x] Remove perspective distortion from PCF and Poisson filters
* [ ] Tune the light frustum for constant pixel density

## How to test

* Turn on shadow mapping and set soft shadow radius to 1.0 (minimum and default)
* Start a game and observe the shadows.
* Shadow edges should appear uniformly sharp, also for distant shadows.
* While jagged at some presets, edges do not have Fresnel lens pattern (see #11358)
